### ISSUE TITLE
[C23] struct member attribute [[deprecated]] ignored on C23 empty initializer

### ISSUE THREAD

**Issue by xBiggs** [llvm/llvm-project#111074]
Observed on `clang 19.1.0`

Compiled with `-std=c23 -Wpedantic`

Consider:
https://godbolt.org/z/dzx1axMhW

```c
struct color {
    [[deprecated]] char r;
    [[deprecated]] char g;
    [[deprecated]] char b;
};

int main(void) {
    [[maybe_unused]] struct color red = {0};
    [[maybe_unused]] struct color green = {};
}
```

The ANSI C zero initializer used on `red` yields a deprecated warning.
The C++ style C23 empty initializer used on `green` does not generate any warning.

```
<source>:8:42: warning: 'r' is deprecated [-Wdeprecated-declarations]
    8 |     [[maybe_unused]] struct color red = {0};
      |                                          ^
<source>:2:7: note: 'r' has been explicitly marked deprecated here
    2 |     [[deprecated]] char r;
      |       ^
1 warning generated.
```

**Comment by shafik** [comment]
That seems like plausible behavior b/c you are explicitly initializing `r` in the `red` case but @AaronBallman would likely have a better feeling on this one.

**Comment by xBiggs** [comment]
Hopefully, you can see how powerful this could be with the example below.

```c
// color.h
#ifndef COLOR_H
#define COLOR_H

enum color { COLOR_RED };

typedef struct Color Color;

struct Color {
  [[deprecated("Private")]] enum color color;
};

extern const Color RED;

enum color GetValue(Color color);

#endif

```

```c
// color.c
#include "color.h"

[[deprecated("Implementation")]] const Color RED = { COLOR_RED };

[[deprecated("Implementation")]] enum color GetColor(Color color) {
  return color.color;
}

```

**Comment by AaronBallman** [comment]
C23 6.7.11p11: ... If an object that has static or thread storage duration is not initialized explicitly, or any object is initialized with an empty initializer, then it is subject to default initialization, which initializes an object as follows: ... if it is an aggregate, every member is initialized (recursively) according to these rules, and any padding is initialized to zero bits ...

But then p22 says: If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate is subject to default initialization.

So if we consider the default initialization to "refer to a name or entity" (6.7.13.5p6), that suggests `{0}` and `{}` should both issue three deprecation diagnostics. Right now, it looks like we only consider it to refer to the name or entity when there's an explicit initializer which is why we only diagnose `{0}` and only with one diagnostic instead of three.

Worth noting that GCC doesn't give any diagnostic for either case: https://godbolt.org/z/rbTjsPb8s

I'm not yet certain what the correct approach is in this case, as the members are not being named anywhere but they are being initialized which is a kind of "reference" in some sense.

**Comment by zygoloid** [comment]
I think that the intent of `[[deprecated]]` is generally to mark things that are planned to be removed. So we should warn in cases that will fail to compile or change meaning when that happens. Thus:

- a designator naming the field should warn
- an implicit initializer for the field should not warn
- an explicit initializer for the field should warn
    - but an idiomatic zero initialization `{0}` should not warn

In the final case an argument could be made for warning if all scalar fields are deprecated, but I don't think it's worth it (and the library author might be planning to add new fields when they remove the old ones).

**Comment by xBiggs** [comment]
> I think that the intent of `[[deprecated]]` is generally to mark things that are planned to be removed. So we should warn in cases that will fail to compile or change meaning when that happens. Thus:
> 
> - a designator naming the field should warn
> - an implicit initializer for the field should not warn
> - an explicit initializer for the field should warn
>     - but an idiomatic zero initialization `{0}` should not warn
> 
> In the final case an argument could be made for warning if all scalar fields are deprecated, but I don't think it's worth it (and the library author might be planning to add new fields when they remove the old ones).

Jens Gustedt of the ISO C working group uses C23's `[[deprecated]]` in the exact same manner as my example.
https://gustedt.wordpress.com/2024/01/06/the-deprecated-attribute-in-c23-does-much-more-than-marking-obsolescence/

**Comment by xBiggs** [comment]
> So if we consider the default initialization to "refer to a name or entity" (6.7.13.5p6), that suggests `{0}` and `{}` should both issue three deprecation diagnostics. Right now, it looks like we only consider it to refer to the name or entity when there's an explicit initializer which is why we only diagnose `{0}` and only with one diagnostic instead of three.

I would be very happy if they both issues three deprecation diagnostics.

**Comment by zygoloid** [comment]
Using `[[deprecated]]` to express something other than deprecation seems like intentionally misleading the compiler, and doing that in general will lead to bad diagnostics.

**Comment by AaronBallman** [comment]
> Using `[[deprecated]]` to express something other than deprecation seems like intentionally misleading the compiler, and doing that in general will lead to bad diagnostics.

100% agreed; that's an abuse of the attribute.

Clang supports a few function attributes that could perhaps be extended to support other entities such as fields (`diagnose_if`, `warning`, and `error` attributes) which might be a better avenue for diagnosing accesses based on user-defined conditions.

**Comment by AaronBallman** [comment]
> I think that the intent of `[[deprecated]]` is generally to mark things that are planned to be removed. So we should warn in cases that will fail to compile or change meaning when that happens. Thus:
> 
>     * a designator naming the field should warn
> 
>     * an implicit initializer for the field should not warn
> 
>     * an explicit initializer for the field should warn
>       
>       * but an idiomatic zero initialization `{0}` should not warn
> 
> 
> In the final case an argument could be made for warning if all scalar fields are deprecated, but I don't think it's worth it (and the library author might be planning to add new fields when they remove the old ones).

Thanks, I actually came to the same model over the weekend based on the same reasoning: the point to the attribute is to alert the programmer that a particular construct will be modified in a potentially breaking way in the future (https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2334.pdf).

As this is the behavior Clang currently implements (https://godbolt.org/z/xxEY9edMz), I think this issue should be closed.

**Comment by erichkeane** [comment]
> > I think that the intent of `[[deprecated]]` is generally to mark things that are planned to be removed. So we should warn in cases that will fail to compile or change meaning when that happens. Thus:
> > ```
> > * a designator naming the field should warn
> > 
> > * an implicit initializer for the field should not warn
> > 
> > * an explicit initializer for the field should warn
> >   
> >   * but an idiomatic zero initialization `{0}` should not warn
> > ```
> >   
> > In the final case an argument could be made for warning if all scalar fields are deprecated, but I don't think it's worth it (and the library author might be planning to add new fields when they remove the old ones).
> 
> Thanks, I actually came to the same model over the weekend based on the same reasoning: the point to the attribute is to alert the programmer that a particular construct will be modified in a potentially breaking way in the future (https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2334.pdf).
> 
> As this is the behavior Clang currently implements (https://godbolt.org/z/xxEY9edMz), I think this issue should be closed.

I agree as well, though note that Clang doesn't implement the sub-point of point 3.  We still diagnose in that case, though I'm less-convinced on that than the others.  I think we should leave this open to discuss/capture the wish to not diagnose on the `idomatic zero init`, which seems reasonably sensible.  I think it is ONLY on a single `0`, else 'removal of a field' ends up initializing a different thing.

**Comment by AaronBallman** [comment]
> > > I think that the intent of `[[deprecated]]` is generally to mark things that are planned to be removed. So we should warn in cases that will fail to compile or change meaning when that happens. Thus:
> > > ```
> > > * a designator naming the field should warn
> > > 
> > > * an implicit initializer for the field should not warn
> > > 
> > > * an explicit initializer for the field should warn
> > >   
> > >   * but an idiomatic zero initialization `{0}` should not warn
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In the final case an argument could be made for warning if all scalar fields are deprecated, but I don't think it's worth it (and the library author might be planning to add new fields when they remove the old ones).
> > 
> > 
> > Thanks, I actually came to the same model over the weekend based on the same reasoning: the point to the attribute is to alert the programmer that a particular construct will be modified in a potentially breaking way in the future (https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2334.pdf).
> > As this is the behavior Clang currently implements (https://godbolt.org/z/xxEY9edMz), I think this issue should be closed.
> 
> I agree as well, though note that Clang doesn't implement the sub-point of point 3. We still diagnose in that case, though I'm less-convinced on that than the others. I think we should leave this open to discuss/capture the wish to not diagnose on the `idomatic zero init`, which seems reasonably sensible. I think it is ONLY on a single `0`, else 'removal of a field' ends up initializing a different thing.

We do handle the sub-point properly. We diagnose the first field if it's deprecated, but no other fields (because with `{0}` there is an explicit initializer for the first field, but implicit for the rest). That's the `struct T t2;` example.

Since we've got agreement amongst maintainers, I'm closing the issue as being by design.

**Comment by gustedt** [comment]
> > Using `[[deprecated]]` to express something other than deprecation seems like intentionally misleading the compiler, and doing that in general will lead to bad diagnostics.
> 
> 100% agreed; that's an abuse of the attribute.
> 
> Clang supports a few function attributes that could perhaps be extended to support other entities such as fields (`diagnose_if`, `warning`, and `error` attributes) which might be a better avenue for diagnosing accesses based on user-defined conditions.

That boils down to the meaning of the English term "deprecated". I am not a native so you may come to a different conclusion than me. But I see that "deprecated" seems also to have the sense of "frowned upon" not only of  "outdated". So I stand with my use case.

(And here the possible "reason" provided as an argument to the attribute has all its value.)

**Comment by gustedt** [comment]
Generally I think that the solution chosen here is the right one. A deprecation of the individual members of a `struct`  should be different from a deprecation of the `struct` as a whole. So default initialization should definitively be allowed and explicit use of a deprecated member should raise eyebrows.

Using the `{0}` initializer is a grey array that should be treated carefully because of backwards compatibility. But since `[[deprecated]]` was not there before C23, I think that the path chosen by llvm is a good one. If people are adding this attribute to existing code they have to be aware that this basically supposes that default initialization of the `struct` has the correct meaning. So if they are using `{  0 }` they should be warned about this and also change their code to use `{}` if they really mean it.
